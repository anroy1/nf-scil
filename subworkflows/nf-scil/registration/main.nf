include { REGISTRATION_ANATTODWI  } from '../../../modules/nf-scil/registration/anattodwi/main'
include { REGISTRATION_ANTS   } from '../../../modules/nf-scil/registration/ants/main'
include { REGISTRATION_EASYREG } from '../../../modules/nf-scil/registration/easyreg/main'

params.run_surgery = false

workflow REGISTRATION {

    // ** The subworkflow requires at least ch_image and ch_ref as inputs to ** //
    // ** properly perform the registration. Supplying a ch_metric will select ** //
    // ** the REGISTRATION_ANATTODWI module meanwhile NOT supplying a ch_metric ** //
    // ** will select the REGISTRATION_ANTS (SyN or SyNQuick) module. Alternatively, ** //
    // ** NOT supplying ch_metric and activating run_surgery flag with select REGISTRATION_EASYREG ** //

    take:
        ch_image                  // channel: [ val(meta), [ image ] ]
        ch_ref                    // channel: [ val(meta), [ ref ] ]
        ch_metric                 // channel: [ val(meta), [ metric ] ], optional
        ch_mask                   // channel: [ val(meta), [ mask ] ], optional
        ch_segmentation           // channel: [ val(meta), [ flo_segmentation ] ], optional
        ch_ref_segmentation       // channel: [ val(meta), [ ref_segmentation ] ], optional


    main:

        image_warped = Channel.empty()
        ref_warped = Channel.empty()
        transfo_image = Channel.empty()
        transfo_trk = Channel.empty()
        out_segmentation = Channel.empty()
        out_ref_segmentation = Channel.empty()

        ch_versions = Channel.empty()

        if ( ch_metric ) {
            // ** Set up input channel ** //
            ch_register =   ch_image.combine(ch_ref, by: 0)
                                    .combine(ch_metric, by: 0)

            // ** Registration using AntsRegistration ** //
            REGISTRATION_ANATTODWI ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_ANATTODWI.out.versions.first())

            // ** Setting outputs ** //
            image_warped = REGISTRATION_ANATTODWI.out.t1_warped
            transfo_image = REGISTRATION_ANATTODWI.out.transfo_image
            transfo_trk = REGISTRATION_ANATTODWI.out.transfo_trk
        }
        else if ( params.run_surgery ) {
            // ** Set up input channel ** //
            ch_register = ch_ref.combine(ch_image, by: 0)
                                .combine(ch_ref_segmentation, by: 0)
                                .combine(ch_segmentation, by: 0)

            // ** Registration using Easyreg ** //
            REGISTRATION_EASYREG ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_EASYREG.out.versions.first())

            // ** Setting outputs ** //
            image_warped = REGISTRATION_EASYREG.out.flo_reg
            transfo_image = REGISTRATION_EASYREG.out.fwd_field
                .map{ it + [[]] }
            transfo_trk = REGISTRATION_EASYREG.out.bak_field
                .map{ [[]] + it }
            ref_warped - REGISTRATION_EASYREG.out.ref_reg

            // ** Setting optional outputs. If segmentations are not provided as inputs, ** //
            // ** easyreg will outputs synthseg segmentations ** //
            out_segmentation = ch_segmentation ? Channel.empty() : REGISTRATION_EASYREG.out.flo_seg
            out_ref_segmentation = ch_ref_segmentation ? Channel.empty() : REGISTRATION_EASYREG.out.ref_seg


        }
        else {
            // ** Set up input channel, input are inverted compared to REGISTRATION_ANATTODWI. ** //
            if ( ch_mask ) {
                ch_register = ch_ref.combine(ch_image, by: 0)
                                    .combine(ch_mask, by: 0)
            }
            else {
                ch_register = ch_ref.combine(ch_image, by: 0)
                                    .map{ it + [[]] }
            }

            // ** Registration using antsRegistrationSyN.sh or antsRegistrationSyNQuick.sh. ** //
            // ** Has to be defined in the config file or else the default (SyN) will be    ** //
            // ** used.                                                                     ** //
            REGISTRATION_ANTS ( ch_register )
            ch_versions = ch_versions.mix(REGISTRATION_ANTS.out.versions.first())

            // ** Setting outputs ** //
            image_warped = REGISTRATION_ANTS.out.image
            transfo_image = REGISTRATION_ANTS.out.transfo_image
            transfo_trk = REGISTRATION_ANTS.out.transfo_trk
        }

    emit:
        image_warped  = image_warped               // channel: [ val(meta), [ image ] ]
        ref_warped = ref_warped                    // channel: [ val(meta), [ ref ] ]
        transfo_image = transfo_image              // channel: [ val(meta), [ warp, affine ] ]
        transfo_trk   = transfo_trk                // channel: [ val(meta), [ inverseAffine, inverseWarp ] ]
        segmentation = out_segmentation            // channel: [ val(meta), [ segmentation ] ]
        ref_segmentation = out_ref_segmentation    // channel: [ val(meta), [ ref_segmentation ] ]


        versions = ch_versions                 // channel: [ versions.yml ]
}
