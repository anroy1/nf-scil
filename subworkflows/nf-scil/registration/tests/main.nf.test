nextflow_workflow {

    name "Test Subworkflow REGISTRATION"
    script "../main.nf"
    workflow "REGISTRATION"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/registration"

    tag "registration"
    tag "registration/anattodwi"
    tag "registration"
    tag "registration/ants"
    tag "registration/easyreg"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "processing.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("registration - antsRegistration") {
        config "./nextflow.config"
        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/mni_masked_2x2x2.nii.gz")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0_mean.nii.gz")
                    ]}
                input[2] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/fa.nii.gz")
                    ]}
                input[3] = []
                input[4] = []
                input[5] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("registration - SyNQuick") {
        config "./nextflow.config"
        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/mni_masked_2x2x2.nii.gz")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0_mean.nii.gz")
                    ]}
                input[2] = []
                input[3] = []
                input[4] = []
                input[5] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("registration - surgery") {
        config "./nextflow_surgery.config"
        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/mni_masked_2x2x2.nii.gz")
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/b0_mean.nii.gz")
                    ]}
                input[2] = []
                input[3] = []
                input[4] = []
                input[5] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
