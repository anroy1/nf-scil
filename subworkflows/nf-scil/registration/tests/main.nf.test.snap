{
    "registration - antsRegistration": {
        "content": [
            "test__t1_warped.nii.gz:md5:header,e327665710b9c0888bf345a10a73c20f,data,1e29233890cb0b9cd0420d581cc77397",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__output1Warp.nii.gz:md5,f896e4fef21f26c84bf834afd7fb5d55",
                    "test__output0GenericAffine.mat:md5,043decee0d131e79a802e3dcc61c22f0"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__output0GenericAffine.mat:md5,043decee0d131e79a802e3dcc61c22f0",
                    "test__output1InverseWarp.nii.gz:md5,62d6287e60dd36ffd477e71061f23077"
                ]
            ],
            [
                "versions.yml:md5,9c255f4c751382be7bf6fe6ffb0d635d"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-13T15:55:40.238221"
    },
    "registration - SyNQuick": {
        "content": [
            "test__warped.nii.gz:md5:header,e327665710b9c0888bf345a10a73c20f,data,61793ed987f5eac9779dd84acc189f9a",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test__output0Warp.nii.gz:md5,61436a721b29633474255640b5e44f7f",
                        "test__output1GenericAffine.mat:md5,3946e4d33c15911a9a1d72374fdbe4df"
                    ]
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test__output0InverseAffine.mat:md5,b30458b37b728d8ce97890b391ad9df9",
                        "test__output1InverseWarp.nii.gz:md5,278d1774561435ff0a3ec239872b74e5"
                    ]
                ]
            ],
            [
                "versions.yml:md5,96376b8bc28640eef565e7f1e258dbb2"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-13T15:55:52.86551"
    }
}